File: LeetCode root folder/Shortest Distance to a Character.cpp
//bfs solution:3
//ms
//Beats
//15.89%
class Solution {
	public:
		vector<int> shortestToChar(string s, char c) {
			unordered_set<int> visited;
			deque<int> q;
			vector<int> ans(s.size(), 0);
			for (int i = 0; i < s.size(); i++) {
				if (s[i] == c) {
					q.push_back(i);
					visited.insert(i);
                    // cout << q[q.size()-1] << endl;
				}
			}


			int time = 0;
			while (q.size() > 0) {
				int l = q.size();
				while (l > 0) {
					int cur = q.front();
					q.pop_front();
					ans[cur] = time;
					if (visited.find(cur-1) == visited.end() && cur > 0) {
                        visited.insert(cur-1);
						q.push_back(cur-1);
					}
					if (visited.find(cur+1) == visited.end() && cur < s.size()-1) {
						visited.insert(cur+1);
						q.push_back(cur+1);
					}
					l--;
				}
				time++;
			}
			return ans;
		}
	};

File: LeetCode root folder/Shortest Distance to a Character.py
# bfs solution:3
# ms
# Beats
# 68.95%
class Solution:
    def shortestToChar(self, s: str, c: str) -> List[int]:
        visited = set()
        q = deque([])
        ans = [0] * len(s)
        for i in range(len(s)):
            if s[i] == c:
                q.append(i)
                visited.add(i)

        time = 0
        while q:
            l = len(q)
            while l:
                cur = q.popleft()
                ans[cur] = time
                if cur - 1 not in visited and cur > 0:
                    visited.add(cur - 1)
                    q.append(cur - 1)
                if cur + 1 not in visited and cur < len(s) - 1:
                    visited.add(cur + 1)
                    q.append(cur + 1)
                l -= 1
            time += 1
        return ans